lib/model/recent_senders.dart

```diff
@@ -0,0 +1,189 @@
+import 'dart:math' as math;
+
+import '../api/model/model.dart';
+
+extension Max<T extends num> on Iterable<T> {
+  /// Finds the maximum number in an [Iterable].
+  ///
+  /// Returns null if the [Iterable] is empty.
+  T? max() => isNotEmpty ? reduce(math.max) : null;
+}
+
+extension PossibleIndex<T extends num> on List<T> {
+  /// The index of [element] in this sorted list.
+  ///
+  /// Uses binary search to find the index, so the list must be sorted,
+  /// otherwise the result is unpredictable.
+  ///
+  /// If [element] is found, returns its index (0-based).
+  ///
+  /// ```dart
+  /// final somePrimes = [11, 13, 17, 23, 27, 29];
+  /// final index = somePrimes.possibleIndexOf(17); // 2
+  /// ```
+  /// If [element] is not found, returns the possible index (1-based)
+  /// with a negative sign, where if the [element] is inserted,
+  /// the order of the list is maintained.
+  ///
+  /// ```dart
+  /// final somePrimes = [11, 13, 17, 23, 27, 29];
+  /// int index = somePrimes.possibleIndexOf(7); // -1
+  /// index = somePrimes.possibleIndexOf(19); // -4
+  /// ```
+  int possibleIndexOf(T element) {
+    int low = 0;
+    int high = length - 1;
+
+    while (high >= low) {
+      int mid = (low + high) ~/ 2;
+      if (element == this[mid]) {
+        return mid;
+      } else if (element < this[mid]) {
+        high = mid - 1;
+      } else {
+        low = mid + 1;
+      }
+    }
+    return -low - 1;
+  }
+}
+
+/// A data structure to keep track of message ids.
+class IdTracker {
```

Comment:
I think `MessageIdTracker` would be a more helpful name for this, since a lot of the nearby code is dealing with user IDs. 🙂

With that name, the dartdoc as written would stand out as redundant/useless, though, wouldn't it…but that could be dealt with by removing it. But if there's anything else that's interesting about this class (other than that it's a data structure that keeps track of message IDs), that could be good to put in the dartdoc.